src/Ng/Jobs/MonitoringProcessorJob.cs

```diff
@@ -115,20 +129,68 @@ private static void SetUserAgentString()
+                    queueLoopExpiredCancellationToken,
+                    cancellationToken)
+                .Token;
+
+            await ParallelAsync.Repeat(() => ProcessPackagesAsync(queueLoopCancellationToken, cancellationToken));
         }
 
-        private async Task ProcessPackagesAsync(CancellationToken token)
+        /// <remarks>
+        /// We are using two separate <see cref="CancellationToken"/>s. 
+        /// One is for cancelling the queue loop (<paramref name="queueLoopCancellationToken"/>).
+        /// The other is for cancelling processing queue messages (<paramref name="queueMessageCancellationToken"/>).
+        /// This is because if the queue loop expires, we don't want to immediately cancel the processing of all messages.
+        /// We want these messages to finish being handled before ending the loop if possible.
+        /// </remarks>
+        private async Task ProcessPackagesAsync(
+            CancellationToken queueLoopCancellationToken, 
+            CancellationToken queueMessageCancellationToken)
         {
-            StorageQueueMessage<PackageValidatorContext> queueMessage;
+            Logger.LogInformation("Beginning fetching queue messages.");
+            while (!queueLoopCancellationToken.IsCancellationRequested)
+            {
+                await HandleQueueMessagesAsync(queueLoopCancellationToken, queueMessageCancellationToken);
+            }
+
+            Logger.LogInformation("Finished fetching queue messages.");
+        }
+
+        private async Task HandleQueueMessagesAsync(
+            CancellationToken queueLoopCancellationToken,
+            CancellationToken queueMessageCancellationToken)
+        {
+            StorageQueueMessage<PackageValidatorContext> queueMessage = null;
             do
             {
-                Logger.LogInformation("Fetching next queue message.");
-                queueMessage = await _queue.GetNextAsync(token);
-                await HandleQueueMessageAsync(queueMessage, token);
-            } while (queueMessage != null);
+                try
+                {
+                    Logger.LogInformation("Fetching next queue message.");
+                    queueMessage = await _queue.GetNextAsync(queueMessageCancellationToken);
+                    await HandleQueueMessageAsync(queueMessage, queueMessageCancellationToken);
+
+                    if (queueMessage == null)
+                    {
+                        Logger.LogInformation(
+                            "Failed to fetch last message or no messages left in queue. Waiting {EmptyQueueDelaySeconds} seconds before polling again.",
+                            _emptyQueueDelay.TotalSeconds);
 
-            Logger.LogInformation("No messages left in queue.");
+                        await Task.Delay(_emptyQueueDelay, queueLoopCancellationToken);
+                    }
+                }
+                catch (Exception e)
+                {
+                    Logger.LogCritical(
```

Comment:
There should be a delay in the exception path IMO

Reply:
Sounds good.